### ISSUE TITLE
Training Strategy, Data Splits, and Hard Negatives for YOLOv11-Nano

### ISSUE THREAD

**Issue by fkaanoz** [ultralytics/ultralytics#22888]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi,
I am quite new to deep learning and model training. I am working on my capstone project mostly on my own, and unfortunately I don’t have anyone around me to ask technical questions. So I would really appreciate any guidance or feedback from the community. 

I know I am asking many questions, and I don’t expect answers to all of them. I wrote everything explicitly as much as I can, because I want to document my considerations clearly, and hopefully this issue can also help others who are working on similar problems. Any guidance or partial answers are more than enough. I appreciate your time and effort.

**General / SPLITTING :** 
I am training a YOLOv11-Nano model to detect drones. In the initial version, I ignored other flying objects such as birds. 
- Total images: 42K+ 
- Number of datasets used: 11 (WOSDeCT, DetFly, UAV-Eagle, the MAV dataset, and others)

I tried three different splitting strategies:
1- Fully random split
2- Dataset-based split
3- Clustering all the images based on bounding-box width–height on a scatter distribution.

After testing for a few epochs:
The random split produced artificially high mAP50, probably because of distribution leakage.
The dataset-based split showed poor generalization for 30 epochs training. mAP50 fluctuates during training.
Because of this, I continued with approach #3, which maintains a kinda balanced width–height distribution between train and validation sets. 

My current split is: Train: 35K and Validation: 7K

----- 

Due to Kaggle’s 12-hour runtime limit, I am training the model section by section, resuming from the last checkpoint each time. For different sections, I also changed optimizer, momentum, and LR.

<img width="2886" height="1404" alt="Image" src="https://github.com/user-attachments/assets/ae7fdee6-d071-4e55-b65d-b2c57bc3c28e" />
(Figure 1)

The graph (Figure1) shows mAP50 and mAP50-95 across training iterations. I divided my work into three sections:

* Section 1
Used on-the-fly augmentations (help of parameters of train method)
Optimizer: SGD, Momentum: 0.937
Later, I noticed some augmentations (especially mosaic) sometimes produced unrealistic images where large drones were represented only by tiny parts.

* Section 2
Disabled all augmentations.
Optimizer: SGD, Momentum: 0.95
This produced more stable performance, but I still observed false positives when evaluating on an totally unrelated dataset. 

* Section 3 
To reduce false positives, I added hard negatives: +1.3K background images to train +0.8K background images to validation Backgrounds were collected from 20+ YouTube videos recorded arround the campus (1 frame per second, 10+ hours ≈ 36K frames). 
Also, I froze the first 4 backbone layers (by setting freeze=4) so I wouldn’t lose feature extraction learned during Section 2. 
Optimizer: AdamW, Momentum: 0.92

----- 

Now I have dozens of question: 

* Issue 1:
After adding hard negatives, mAP50 and mAP50-95 slightly increased, but I noticed more confidence fluctuation during inference. For example, confidence now varies between 0.50–0.75, whereas in Section 2 it was more stable (0.65–0.75). (Note that I currently have about 3% hard negatives in the training set. )

How can I reduce these confidence fluctuations? Should I add more hard negatives? Is this expected behavior after introducing difficult backgrounds even though the increase in mAP50 results?

* Issue 2:
I am a little bit suspicious about my train–valid split because mAP50 is already somewhat high in the first epoch. The increase over epochs looks natural (not like a sudden spike) so I continued with this split. 

Does the pattern in the graphs suggest any data leakage? Or do these early-epoch results seem normal for this type of dataset (in terms of size) ? 

* Issue 3: 
Due to Kaggle’s 12-hour runtime limit, I am training the model section by section, resuming from the best.pt of the last section each time. During different sections, I also changed optimizer, momentum, and LR. 

Can this approach cause problems? Is multi-stage training with changing hyperparameters safe? Could optimizer switching harm convergence?

* Issue 4: 
For the final model, I am preparing a refined dataset. I use the width–height scatter plot to divide bounding boxes into 5 diagonal-size based regions. 
Based on the region, I apply specific offline augmentations: 
Ex: Large bounding boxes: noise, blur, etc. 
Ex: 75–150 px diagonal bounding boxes: mosaic 
Ex: Others: different and a little bit light augmentations like noise, saturation, etc. 
 
Can you evaluate this offline pre-augmentation approach compared to on-the-fly augmentation? Which one is more appropriate for my project?

* Issue 5: For output of the Section 2, I noticed that the model becomes sensitive to objects that resemble the shape of the drone’s silhouette. For example, distant building columns that visually mimic the π symbol. These structures are sometimes detected as drones, or objects when they appear against the sky background. 
At first, I interpreted this as a potentially good sign, because it suggests some form of geometric awareness rather than memorizing only. However, I am not sure if this is expected or desirable behavior. Is this a good sign?

keywords: drone detection, single class drone detection, hard negatives for single class detection, train-valid split, data leakage

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fkaanoz, thank you for your detailed question and for using Ultralytics YOLO 🚀! This is an automated response to help you get oriented quickly — an Ultralytics engineer will also review your issue and follow up as soon as possible 😊  

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. A lot of common questions around training strategies, data splits, augmentation, and validation behavior are covered there and may help you refine your experiments.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

Since your post is a custom training ❓ Question with multiple experimental setups, please include as much concrete information as possible to help others reproduce or reason about your results, such as:
- Exact training commands (including key arguments like `data=`, `model=`, `imgsz=`, `epochs=`, `batch=`, `optimizer=`, `freeze=`, etc.)
- A short summary of your dataset structure and how you created the different splits
- Representative sample images (including your drone class and some of your hard negatives)
- Relevant training logs and the specific mAP metrics/curves you are referring to for each “section” of training

Also, please verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), as they address many topics related to:
- Splitting strategies and avoiding data leakage
- Augmentation choices (including mosaic and other transforms)
- Hyperparameter tuning and optimizer configuration
- Multi-stage training and convergence behavior

Join the Ultralytics community where it suits you best to discuss training strategies like yours in more depth:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- For longer-form, technical discussions, check out [Discourse](https://community.ultralytics.com/)  
- Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share your findings and learn from similar projects  

## Upgrade

Before diving deeper, please ensure you are on the latest `ultralytics` package so that your YOLO11-Nano experiments benefit from the newest features and fixes. Upgrade in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

These can be especially useful for longer training runs or when experimenting with different strategies (e.g., hard negatives, multi-stage training, or alternative augmentations) beyond Kaggle’s time limits.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the very clear write-up – this is a nice, thoughtful approach for a capstone project, so you’re on a good track already.

Issue 1 – hard negatives and confidence:  
Adding hard negatives generally pushes the model to be more cautious, so a slight drop and wider spread in confidences (e.g., 0.5–0.75 instead of 0.65–0.75) is expected and not a problem by itself, especially if mAP improved. Rather than trying to “stabilize” the raw confidence values, tune your deployment `conf` and `iou` thresholds based on the validation set and keep adding diverse backgrounds similar to where you see false positives; a background fraction of a few to ~10% is perfectly reasonable.

Issue 2 – split and potential leakage:  
With a pretrained YOLO11n and 42k images of a single, visually distinctive class, seeing a relatively high mAP already in the first epoch is normal and does not automatically indicate leakage. The key leakage risks here are near-duplicate frames from the same video or overlapping source datasets ending up in both train and val; if you ensure that sequences and exact images are strictly disjoint across splits, your current width–height–balanced split is fine. General guidance on splits and avoiding leakage is summarized in the data preprocessing guide at [Data preprocessing and dataset splitting](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

Issue 3 – sectioned training and changing optimizers:  
Multi-stage training is safe as long as you’re clear on whether you’re resuming the full training state or just reusing weights. If you truly want to “switch optimizer,” treat each section as a new run initialized from the previous `best.pt`, for example: `yolo train model=best.pt ...` with your new `optimizer`, `lr0`, and `momentum`; this discards the old optimizer state and is essentially standard fine-tuning. Using wildly different learning rates between sections is the main thing that could hurt convergence, so keep LR changes moderate and watch the loss/mAP curves.

Issue 4 – offline vs on-the-fly augmentation:  
In general, on-the-fly augmentation via YOLO’s training config is preferred: it keeps the dataset small, produces effectively infinite variations, and makes it easy to adjust policies without regenerating data. Offline, size-dependent augmentation as you describe can work, but it locks you into a fixed set of synthetic images and can over-emphasize certain patterns if you duplicate them heavily. For your drone task I’d keep augmentations online, mostly moderate color and geometric jitter, and if mosaic creates unrealistic partial drones, reduce its probability (`mosaic` < 1.0) and/or turn it off earlier in training with `close_mosaic`, as suggested in the [YOLO data augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/). You’ll get most of the benefit of scale-aware training just from `scale`, `translate`, and varied object sizes already present in your data.

Issue 5 – π-shaped buildings being detected as drones:  
This is very typical behavior: the model is latching onto the most discriminative local shape it has seen for “drone,” so anything with a similar silhouette against sky gets some probability mass. It’s not a bad sign, just evidence that your negative/background coverage for those structures is still limited. The right response is exactly what you’re already moving toward: include those confusing structures as hard negatives (or even as a separate “non-drone flying/structure” class if that’s useful), and continue training so the model learns to use more context than just the silhouette.

If you iterate along these lines and keep the split clean and augmentations simple and on-the-fly, you should be in a good place; feel free to share a small subset plus your exact `yolo train` command if you’d like more targeted suggestions on the training config.

**Comment by fkaanoz** [comment]
I cannot thank enough, Mr. Jocher. Your help will be incredibly valuable.

**Comment by glenn-jocher** [comment]
Glad it helped, and nice work laying everything out so clearly — as you keep iterating, feel free to follow up here with specific experiments or new findings you’d like feedback on.
